README.md

```diff
@@ -1,84 +1,168 @@
+```
+ls\n
 
-A good path name should be decipherable -- meaning that if some machine or developer -- who has no idea about your protocol -- encounters the path string, they should be able to look it up and resolve how to use it.
+# in hex (note the varint prefix = 3)
+0x036c730a
+```
 
-An example of a good path name is:
+So a remote side asking for a protocol listing would look like this:
 
+```sh
+# request
+<multistream-header-for-multistream-select>
+ls\n
+
+# response
+<varint-total-response-size-in-bytes><varint-number-of-protocols>
+<multicodec-of-available-protocol>
+<multicodec-of-available-protocol>
+<multicodec-of-available-protocol>
+...
 ```
-/bittorrent.org/1.0
+
+For example
+
+```sh
+# send request
+> /ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+> ls
+
+# get response
+< /ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/1.0.0
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-bitswap/0.4.3
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-bitswap/1.0.0
+
+# send selection, upgrade connection, and start protocol traffic
+> /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+> <ipfs-dht-request-0>
+> <ipfs-dht-request-1>
+> ...
+
+# receive selection, and upgraded protocol traffic.
+< /ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+< <ipfs-dht-response-0>
+< <ipfs-dht-response-1>
+< ...
 ```
 
-An example of a _great_ path name is:
+### Example
 
 ```
-/ipfs/Qmaa4Rw81a3a1VEx4LxB7HADUAXvZFhCoRdBzsMZyZmqHD/ipfs.protocol
-/http/w3id.org/ipfs/ipfs-1.1.0.json
+# greeting
+> /http/multiproto.io/multistream-select/1.0
+< /http/multiproto.io/multistream-select/1.0
+
+# list available protocols
+> /http/multiproto.io/multistream-select/1.0
+> ls
+< /http/google.com/spdy/3
+< /http/w3c.org/http/1.1
+< /http/w3c.org/http/2
+< /http/bittorrent.org/1.2
+< /http/git-scm.org/1.2
+< /http/ipfs.io/exchange/bitswap/1
+< /http/ipfs.io/routing/dht/2.0.2
+< /http/ipfs.io/network/relay/0.5.2
+
+# select protocol
+> /http/multiproto.io/multistream-select/1.0
+> ls
+> /http/w3id.org/http/1.1
+> GET / HTTP/1.1
+>
+< /http/w3id.org/http/1.1
+< HTTP/1.1 200 OK
+< Content-Type: text/html; charset=UTF-8
+< Content-Length: 12
+<
+< Hello World
 ```
 
-These path names happen to be resolvable -- not just in a "multistream muxer" -- but in the internet as a whole (provided the program (or OS) knows how to use the `/ipfs` and `/http` protocols).
+### ls example in detail
 
-## Implementations
+```
+> <varint-length>ls<newline>
+< <varint-total-response-size-in-bytes><varint-number-of-protocols><newline>
```

Comment:
For record, `MOAR varints` translates to another varint that should be added on the beginning of line 158 that tells the byte size of the following two varints + newline